alacritty_terminal/src/tty/unix.rs

```diff
@@ -121,17 +118,38 @@ impl Pty {
     }
 }
 
-#[cfg(target_os = "macos")]
-fn default_shell(pw: &Passwd<'_>) -> Program {
-    let shell_name = pw.shell.rsplit('/').next().unwrap();
-    let argv = vec![String::from("-c"), format!("exec -a -{} {}", shell_name, pw.shell)];
+fn shell_command(shell_config: Option<&Program>, pw: &Passwd<'_>) -> Command {
+    // Look for a shell in the alacritty config first, then SHELL, then in passwd.
+    let shell = shell_config
+        .map(|prog| prog.program().to_owned())
+        .or_else(|| env::var("SHELL").ok())
+        .unwrap_or_else(|| pw.shell.to_owned());
 
-    Program::WithArgs { program: "/bin/bash".to_owned(), args: argv }
-}
+    #[cfg(not(target_os = "macos"))]
+    let mut command = Command::new(shell);
+
+    // On macOS, unless the user set the `shell` config in alacritty.yml, use the `login` command to
+    // launch the shell so programs like `w`, `who`, and `logname` work.
+    #[cfg(target_os = "macos")]
```

Comment:
This whole thing should not clutter up the codebase for people just trying to figure out how this works on sensible linux/bsd systems. There's a good reason why the macos stuff was extracted into a separate function.

Reply:
Sure, I'll make separate versions of the function.